File: recyclerview/src/main/java/com/example/recyclerview/MainActivity.java
package com.example.recyclerview;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    interface ListItemActionListener{
        void onItemClicked(int number);
    }

    private static final List<Integer> itemList;
    static {
        itemList = new ArrayList<>();
        for (int i = 0; i < 150; i++) {
            itemList.add(i);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView recyclerView = findViewById(R.id.itemList);
        recyclerView.setAdapter(new ItemListAdapter(itemList, new ListItemActionListener() {
            @Override
            public void onItemClicked(int number) {
                Log.d("LISTENER", String.valueOf(number));
            }
        }));
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
    }

    static class ItemListAdapter extends RecyclerView.Adapter<ItemListAdapter.ItemViewHolder> {

        private List<Integer> items;
        private ListItemActionListener listItemActionListener;

        public ItemListAdapter(List<Integer> items, ListItemActionListener listItemActionListener) {
            this.items = items;
            this.listItemActionListener = listItemActionListener;
        }

        public void setItems(List<Integer> items) {
            this.items = items;
            notifyDataSetChanged();
        }

        @NonNull
        @Override
        public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_number_title, parent, false);
            return new ItemViewHolder(view, listItemActionListener);
        }

        @Override
        public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
            holder.bind(items.get(position));
        }

        @Override
        public int getItemCount() {
            return items != null ? items.size() : 0;
        }

        static class ItemViewHolder extends RecyclerView.ViewHolder {

            private ImageView imageView;
            private TextView textView;
            private ListItemActionListener listItemActionListener;

            public ItemViewHolder(@NonNull View itemView, ListItemActionListener listItemActionListener) {
                super(itemView);
                imageView = itemView.findViewById(R.id.viewIcon);
                textView = itemView.findViewById(R.id.textView);
                this.listItemActionListener = listItemActionListener;
            }

            public void bind(final int number) {
                itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (listItemActionListener != null){
                            listItemActionListener.onItemClicked(number);
                        }
                    }
                });
                textView.setText(String.valueOf(number));
            }
        }
    }
}

File: dbexample/src/main/java/com/example/dbexample/MainActivity.kt
package com.example.dbexample

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton

class MainActivity : AppCompatActivity() {

    private lateinit var userAdapter: UserListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val recyclerView = findViewById<RecyclerView>(R.id.itemList)
        userAdapter = UserListAdapter()
        recyclerView.apply {
            adapter = userAdapter
            layoutManager = LinearLayoutManager(this@MainActivity)
        }
        findViewById<FloatingActionButton>(R.id.fabBtn).setOnClickListener {
            startActivity(DataActivity.newIntent(this))
        }
    }

    override fun onResume() {
        super.onResume()
        userAdapter.updateItem(getUsersFromDB())
    }

    private fun getUsersFromDB(): List<UserItem> {
        val cursor = (applicationContext as App)
                .dbHelper
                .readableDatabase
                .query("users", null, null, null, null, null, null)
        if (cursor != null) {
            val idIndex = cursor.getColumnIndex("id")
            val nameIndex = cursor.getColumnIndex("name")
            val list = mutableListOf<UserItem>()
            while (cursor.moveToNext()) {
                list.add(UserItem(cursor.getInt(idIndex), cursor.getString(nameIndex)))
            }
            cursor.close()
            return list
        }
        return emptyList()
    }


    class UserListAdapter() : RecyclerView.Adapter<UserListAdapter.ItemViewHolder>() {
        private val usersList = mutableListOf<UserItem>()
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.user_item, parent, false)
            return ItemViewHolder(view)
        }

        override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
            holder.bind(usersList[position])
        }

        override fun getItemCount(): Int {
            return usersList.size
        }

        fun updateItem(list: List<UserItem>) {
            usersList.apply {
                clear()
                addAll(list)
            }
            notifyDataSetChanged()
        }

        class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

            private val textView = itemView.findViewById<TextView>(R.id.userItem)

            fun bind(userItem: UserItem) {
                textView.text = userItem.name
            }
        }
    }
}